link/kprobe.go

```diff
@@ -0,0 +1,233 @@
+// of the given kernel symbol.
+func kprobe(symbol string, ret bool) (*traceEvent, error) {
+
+	if symbol == "" {
+		return nil, errors.New("symbol name cannot be empty")
+	}
+
+	// Use kprobe PMU if the kernel has it available.
+	tp, err := pmuKprobe(symbol, ret)
+	if err == nil {
+		return tp, nil
+	}
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("creating PMU using sysfs: %w", err)
+	}
+
+	// Use tracefs if kprobe PMU is missing.
+	tp, err = tracefsKprobe(traceGroup, symbol, ret)
+	if err != nil {
+		return nil, fmt.Errorf("creating trace event '%s' in tracefs: %w", symbol, err)
+	}
+
+	return tp, nil
+}
+
+// pmuKprobe returns a TraceEvent to be created using PMU (sysfs).
+func pmuKprobe(symbol string, ret bool) (*traceEvent, error) {
+	et, err := getPMUEventType("kprobe")
+	if err != nil {
+		return nil, err
+	}
+	return &traceEvent{
+		pmuID:     et,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe,
+		finalizer: nil, // no resources are created
+	}, nil
+}
+
+// tracefsKprobe returns a TraceEvent created by writing an entry to
+// <tracefs>/kprobe_events. On Close(), the entry is removed.
+// Repeated calls with the same arguments will yield TraceEvents without
+// finalizers, making sure only one removal attempt against
+// <tracefs>/kprobe_events is made during program teardown.
+func tracefsKprobe(group, symbol string, ret bool) (*traceEvent, error) {
+
+	te := traceEvent{
+		group:     group,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe, // kernel only allows kprobe programs to attach
+		finalizer: nil,
+	}
+
+	// Before attempting to create a trace event through tracefs,
+	// check if an event with the same name already exists.
+	tid, err := getTraceEventID(group, kprobeTraceEventName(symbol, ret))
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		// Return errors other than ErrNotExist to the caller.
+		return nil, fmt.Errorf("getting trace event id: %w", err)
+	}
+	if err == nil {
+		// Kprobe trace event already exists, return a TraceEvent to the caller.
+		// Since this kprobe event was created by a prior call to this function,
+		// or even by another tool or process, it should not be removed by Close().
+		// Leave the finalizer empty.
+		te.finalizer = nil
+		te.tracefsID = tid
+		return &te, nil
+	}
+
+	// Kprobe trace event doesn't exist yet, create it using tracefs.
+	if err := createTraceFSKprobeEvent(group, symbol, ret); err != nil {
+		return nil, fmt.Errorf("creating kprobe event on tracefs: %w", err)
+	}
+
+	// Get the newly-created trace event's id.
+	tid, err = getTraceEventID(group, kprobeTraceEventName(symbol, ret))
+	if err != nil {
+		return nil, fmt.Errorf("getting trace event id: %w", err)
+	}
+
+	// An entry was written to <tracefs>/kprobe_events in this call,
+	// so set a finalizer that undoes the operation when this TraceEvent
+	// is closed.
+	te.finalizer = func() error { return closeTraceFSKprobeEvent(group, symbol, ret) }
```

Comment:
Correct, `a.Close()` kills `b`, but only on kernels 4.15 and lower, since the first call writes a global entry to tracefs. My rationale was that successive `Kprobe()` calls would get their `Close()`s deferred in order, which would automatically result in `a` being closed last, since `defer` is a stack. I'm struggling to document this trap without going to deep into the weeds, but maybe briefly mentioning PMU vs. kprobe_events would be useful. (as well as mentioning that on 4.15 and earlier, kprobe events are global)

The finalizer approach felt like the most straightforward to me. The alternative would be a refcounting pool that closes the kprobe event when the last handle is closed, _but only if the current process is the one that wrote the record initially_. Also consider that:
- the caller might want to avoid closing a kprobe_event handle on purpose after pinning a program and perf ring
- other processes can create these events as well, so any pre-existing handle should never be taken ownership over

Reply:
Will remove the finalizer system, see https://github.com/cilium/ebpf/pull/265#discussion_r602519878.